### PR TITLE
[Module interface] Use features in module interface generation.

### PR DESCRIPTION
**Pull request by DougGregor**
When generating a module interface, emit `#if` around any declarations
that are tied to specific, named language features. This allows module
interfaces to be processed by older Swift compilers that do not
support these newer features, such as async/await or actors.

The amount of effort required to correctly handle a new kind of
feature varies somewhat drastically based on the feature itself. The
"simple" case is where a particular declaration can only exist if a
feature is available. For example, and `async` declaration is fairly
easy to handle; a `@_marker` protocol's conformances are not.

Fixes rdar://73326633.

Design credit to @beccadax, bugs are my own.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: lib/AST/ASTPrinter.cpp [hunk 1]
```diff
@@ -2373,6 +2392,143 @@ void PrintAST::printExtension(ExtensionDecl *decl) {
...
+  }
+
+  if (auto ext = dyn_cast<ExtensionDecl>(decl)) {
+    if (auto classDecl = ext->getSelfClassDecl())
+      if (classDecl->isActor())
+        return true;
+  }
+
+  return false;
+}
```
Comment (beccadax) [inline lib/AST/ASTPrinter.cpp:2454]: Unless I'm mistaken, all three of these probably ought to walk into any parameter or result types and look for the feature being used there, too. For instance, if you write:

```
public func runAsync(_ fn: () async -> Void) { ... }
```

That probably ought to be guarded by `#if $AsyncAwait`, even though `runAsync(_:)` itself isn't an async function. Similarly:

```
public func transact(with account: BankAccount)
````

Probably ought to be guarded by `#if $Actors` if `BankAccount` is an actor class, even though `transact(with:)` isn't itself an actor class.

File: lib/AST/ASTPrinter.cpp [hunk 2]
```diff
@@ -2373,6 +2392,143 @@ void PrintAST::printExtension(ExtensionDecl *decl) {
...
+
+/// Determine the set of "new" features used on a given declaration.
+///
+/// Note: right now, all features we check for are "new". At some point, we'll
+/// want a baseline version.
+static std::vector<Feature> getFeaturesUsed(Decl *decl) {
+  std::vector<Feature> features;
+  
+  // Only type- and module-scope declarations have any features to speak of.
```
Comment (beccadax) [inline lib/AST/ASTPrinter.cpp:2463]: Note that, by my expanded definition of "having a feature", this comment is no longer true.

**Review by beccadax — CHANGES_REQUESTED** [review]

File: lib/AST/ASTPrinter.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/AST/ASTPrinter.cpp:2454]: Good point!

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: lib/AST/ASTPrinter.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/AST/ASTPrinter.cpp]: I went ahead and did a bunch of this checking. It is not comprehensive, I'm sure, although we could probably do some additional meta programming to make it less boilerplate-y and more comprehensive.

File: lib/AST/ASTPrinter.cpp (hunk 2 above)
Comment (DougGregor) [reply lib/AST/ASTPrinter.cpp]: Right, I've zapped this. Thanks!

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by davezarzycki** [comment]
Hi @DougGregor – The test is broken and confusing. Superficially, it needs `REQUIRES: concurrency` but the comments in the test make it sound like we're testing the opposite: "Make sure we can parse the file without concurrency enabled". What's going on?

Also, FYI: `-emit-module-interface-path` supports the traditional Unix stdout trick `-` if you want to avoid needless temp files (which on Apple platforms, aren't as fast as they could/should be).

**Pull request merged**